hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/keyvalue/KvMetadataResolver.java

```diff
@@ -81,4 +85,75 @@ static void maybeAddDefaultField(
             tableFields.add(new MapTableField(fieldName, type, true, QueryPath.create(fieldName)));
         }
     }
+
+    /**
+     * If {@code __key}/{@code this} is the only key/value field and has a custom type with
+     * nonnull metadata, resolve the schema ID from the type. Otherwise, return {@code orElse()}.
+     */
+    static <T> T getSchemaId(
+            Map<QueryPath, MappingField> fields,
+            Function<String, T> resolveFromType,
+            Supplier<T> orElse
+    ) {
+        return flatMap(fields, type -> {
+            String metadata = type.getObjectTypeMetadata();
+            return metadata != null ? resolveFromType.apply(metadata) : orElse.get();
+        }, orElse);
+    }
+
+    /**
+     * If {@code __key}/{@code this} is the only field and has a custom type,
+     * return type fields. Otherwise, return mapping fields without {@code __key} or {@code this}.
+     */
+    static Stream<Field> getFields(Map<QueryPath, MappingField> fields) {
+        return flatMap(fields, type -> type.getObjectFields().stream().map(Field::new),
+                () -> fields.entrySet().stream().filter(e -> !e.getKey().isTopLevel()).map(Field::new));
+    }
+
+    /**
+     * If {@code __key}/{@code this} is the only field and has a custom type,
+     * return {@code typeMapper(fieldType)}. Otherwise, return {@code orElse()}.
+     */
+    static <T> T flatMap(
```

Comment:
name of `flatMap` method is misleading. It does not have much to do with flattening anything and does significantly more that flatMap from streams or optional.
what the meanig of this operation in the problem domain? javadoc mostly translates Java code to English so it is not much help.

Reply:
the same applies to javadoc of other new methods in this class

Reply:
I modified `flatMap()` and `getSchemaId()` to return `Optional`s. They no longer expect any `Function` or `Supplier`, and the caller is doing `.map().orElseGet()` instead. I also renamed the methods to reflect this change.

Reply:
I don't see that in b79e9b64ac19d5b130867985030108674672fd63